### PR TITLE
added structs comparison support

### PR DESCRIPTION
**Pull request by kashifkhan0771**
Added the support to compare two objects of same struct and return the results with old and new value

## Summary by CodeRabbit

- **New Features**
	- Introduced a new package for comparing struct instances, enabling users to easily identify changes between old and new data.
	- Added a `Result` structure to detail differences in field values.

- **Tests**
	- Implemented test cases to validate the functionality of the new comparison feature, ensuring accuracy in detecting changes between struct instances.
	- Added tests for both simple and complex struct comparisons to verify expected outputs.

### CODE REVIEW

File: structs/structs.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+				fieldName = tag
+			}
+
+			if !reflect.DeepEqual(oldFieldValue.Interface(), newFieldValue.Interface()) {
+				comparedResults = append(comparedResults, Result{
+					FieldName: fieldName,
+					OldValue:  oldFieldValue.Interface(),
+					NewValue:  newFieldValue.Interface(),
+				})
+			}
```
[the remark this answers is not in the document]
Comment (kashifkhan0771) [reply structs/structs.go:52]: You can ignore this for now.

File: structs/structs_test.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+			if (err != nil) != tt.wantErr {
+				t.Errorf("CompareStructs() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("CompareStructs() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
```
[the remark this answers is not in the document]
Comment (kashifkhan0771) [reply structs/structs_test.go:88]: Will add pointer field support later.

[the remark this answers is not in the document]
Comment (kashifkhan0771) [reply structs/structs_test.go:88]: Yeah sure

**Comment by shahzadhaider1** [comment]
retest

**Comment by shahzadhaider1** [comment]
Please rebase with main so it run all the checks

File: structs/structs.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+*/
+func CompareStructs(old, new interface{}) ([]Result, error) {
+	if reflect.TypeOf(old) != reflect.TypeOf(new) {
+		return nil, fmt.Errorf("both structs must be of the same type")
+	}
+
+	oldValue := reflect.ValueOf(old)
+	newValue := reflect.ValueOf(new)
+
+	var comparedResults = make([]Result, 0)
```
Comment (shahzadhaider1) [inline structs/structs.go:27]: We should initialize the slice with a capacity, possibly using `make([]Result, 0, oldValue.NumField())`. This avoids unnecessary memory allocations when appending items later, which improves performance.

File: structs/structs.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+CompareStructs compares two struct instances of the same type
+and returns a list of results with the old and new values of each field tagged with `updateable`.
+*/
+func CompareStructs(old, new interface{}) ([]Result, error) {
+	if reflect.TypeOf(old) != reflect.TypeOf(new) {
+		return nil, fmt.Errorf("both structs must be of the same type")
+	}
+
+	oldValue := reflect.ValueOf(old)
+	newValue := reflect.ValueOf(new)
```
Comment (shahzadhaider1) [inline structs/structs.go:25]: what about we also make sure that both old and new are actually structs before calling reflection functions using Kind()?

Comment (shahzadhaider1) [reply structs/structs.go:25]: Do we also need to check for possible pointers to structs and dereference them?
```
if oldValue.Kind() == reflect.Ptr {
		oldValue = oldValue.Elem()
	}
```

File: structs/structs_test.go (hunk 2 above)
Comment (shahzadhaider1) [inline structs/structs_test.go:88]: please add tests for:
- pass anything but struct
- pass pointer to struct

**Review by shahzadhaider1 — CHANGES_REQUESTED** [review]

**Comment by shahzadhaider1** [comment]
Please rebase the branch as well as this branch is out-of-date with the base branch

File: structs/structs.go (hunk 4 above)
Comment (kashifkhan0771) [reply structs/structs.go:25]: Pointers I am thinking to implement later on. For now just simple types.

File: structs/structs.go (hunk 3 above)
Comment (kashifkhan0771) [reply structs/structs.go]: done

**Review by shahzadhaider1 — APPROVED** [review]
LGTM

**Pull request merged**